dipy/denoise/shift_twist_convolution.pyx

```diff
@@ -0,0 +1,190 @@
+    return output_sh
+    
+def convolve_sf(odfs_sf, kernel, test_mode=False, num_threads=None):
+    """ Perform the shift-twist convolution with the ODF data and 
+    the lookup-table of the kernel.
+
+    Parameters
+    ----------
+    odfs : array of double
+        The ODF data sampled on a sphere
+    lut : array of double
+        The 5D lookup table
+    test_mode : boolean
+        Reduced convolution in one direction only for testing
+
+    Returns
+    -------
+    output : array of double
+        The ODF data after convolution enhancement
+    """
+    # perform the convolution
+    output = perform_convolution(odfs_sf, 
+                                 kernel.get_lookup_table(),
+                                 test_mode,
+                                 num_threads)
+    return output
+    
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+@cython.cdivision(True)
+cdef double [:, :, :, ::1] perform_convolution (double [:, :, :, ::1] odfs, 
+                                                double [:, :, :, :, ::1] lut,
+                                                cnp.npy_intp test_mode,
+                                                num_threads=None):
+    """ Perform the shift-twist convolution with the ODF data 
+    and the lookup-table of the kernel.
+
+    Parameters
+    ----------
+    odfs : array of double
+        The ODF data sampled on a sphere
+    lut : array of double
+        The 5D lookup table
+    test_mode : boolean
+        Reduced convolution in one direction only for testing
+        
+    Returns
+    -------
+    output : array of double
+        The ODF data after convolution enhancement
+    """
+        
+    cdef:
+        double [:, :, :, ::1] output = np.array(odfs, copy=True)
+        cnp.npy_intp OR1 = lut.shape[0]
+        cnp.npy_intp OR2 = lut.shape[1]
+        cnp.npy_intp N = lut.shape[2]
+        cnp.npy_intp hn = (N - 1) / 2
+        double [:, :, :, :] totalval
+        double [:, :, :, :] voxcount
+        cnp.npy_intp nx = odfs.shape[0]
+        cnp.npy_intp ny = odfs.shape[1]
+        cnp.npy_intp nz = odfs.shape[2]
+        cnp.npy_intp threads_to_use = -1
+        cnp.npy_intp all_cores = openmp.omp_get_num_procs()
+        cnp.npy_intp corient, orient, cx, cy, cz, x, y, z
+        cnp.npy_intp expectedvox
+        cnp.npy_intp edgeNormalization = True
+
+    if num_threads is not None:
+        threads_to_use = num_threads
+    else:
+        threads_to_use = all_cores
+
+    if have_openmp:
+        openmp.omp_set_dynamic(0)
+        openmp.omp_set_num_threads(threads_to_use)
+    
+    if test_mode:
+        edgeNormalization = False
+        OR2 = 1
+
+    # expected number of voxels in kernel
+    totalval = np.zeros((OR1, nx, ny, nz))
+    voxcount = np.zeros((OR1, nx, ny, nz))
+    expectedvox = nx * ny * nz
+
+    with nogil:
+
+        # loop over ODFs cx,cy,cz,corient --> y and v
+        for corient in prange(OR1, schedule='guided'):
```

Comment:
That what nlmeans is using also, which supports stuff like nogil options and num_threads http://docs.cython.org/src/userguide/parallelism.html but I don't know how it plays with the openmp setup that is here. Elef changed the function recently because of threading stuff, so they might be just doing that behind the scenes.